src/JustEat.StatsD/SocketTransport.cs

```diff
@@ -9,23 +9,27 @@ namespace JustEat.StatsD
 {
     /// <summary>
     /// A class representing an implementation of <see cref="IStatsDTransport"/>
-    /// that uses UDP and pools sockets. This class cannot be inherited.
+    /// that uses UDP or IP sockets and pools them.
+    /// This class cannot be inherited.
     /// </summary>
-    public sealed class UdpTransport : IStatsDTransport, IStatsDBufferedTransport, IDisposable
+    public sealed class SocketTransport : IStatsDTransport, IStatsDBufferedTransport, IDisposable
     {
         private ConnectedSocketPool _pool;
         private readonly IPEndPointSource _endpointSource;
+        private readonly SocketProtocol _socketProtocol;
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="UdpTransport"/> class.
+        /// Initializes a new instance of the <see cref="SocketTransport"/> class.
         /// </summary>
         /// <param name="endPointSource">The <see cref="IPEndPointSource"/> to use.</param>
+        /// <param name="socketProtocol">Udp or Ip sockets</param>
         /// <exception cref="ArgumentNullException">
         /// <paramref name="endPointSource"/> is <see langword="null"/>.
         /// </exception>
-        public UdpTransport(IPEndPointSource endPointSource)
+        public SocketTransport(IPEndPointSource endPointSource, SocketProtocol socketProtocol)
         {
             _endpointSource = endPointSource ?? throw new ArgumentNullException(nameof(endPointSource));
+            _socketProtocol = socketProtocol;
```

Comment:
Should we validate the value is valid here, rather than waiting for it to be thrown later when used to create a socket?

Reply:
IMHO not. It's customary to always have a `default` case on a switch on an enum so that it's exhaustive, and never returns null in a quiet failure. You would have to try hard to get an enum value that's not one of the two valid ones.

Reply:
We're validating it in `SocketFactory` though - it's better to throw an error upfront on construction by validating the parameters, than leaving it until later.

Reply:
Amen to that @AnthonySteele. One downside is that `Enum.IsDefined` allocates, although if this class is pooled then it doesn't matter. They aren't going to make runtime changes to remove the allocation (Like they did with `Enum.GetHastCode` in .NET Core 2.0), so we'd need to wait on the enum generic type constraint to land in C# and for a generic version of `Enum.IsDefined` to land.

Reply:
I think based on the use case, the allocation isn't really a concern as you wouldn't expect lots of transports to be being created.